walk/git.go

```diff
@@ -1,169 +1,98 @@
 package walk
 
 import (
+	"bufio"
 	"context"
 	"fmt"
 	"io"
-	"io/fs"
 	"os"
+	"os/exec"
 	"path/filepath"
-	"runtime"
 	"strings"
 
 	"github.com/charmbracelet/log"
-	"github.com/go-git/go-git/v5"
-	"github.com/go-git/go-git/v5/plumbing/filemode"
 	"github.com/numtide/treefmt/stats"
 	"golang.org/x/sync/errgroup"
 )
 
 type GitReader struct {
-	root      string
-	path      string
-	stats     *stats.Stats
-	batchSize int
+	root string
+	path string
 
-	log  *log.Logger
-	repo *git.Repository
+	log   *log.Logger
+	stats *stats.Stats
 
-	filesCh chan *File
-
-	eg *errgroup.Group
+	eg      *errgroup.Group
+	scanner *bufio.Scanner
 }
 
-func (g *GitReader) process() error {
+func (g *GitReader) Read(ctx context.Context, files []*File) (n int, err error) {
+	// ensure we record how many files we traversed
 	defer func() {
-		close(g.filesCh)
+		g.stats.Add(stats.Traversed, int32(n))
 	}()
 
-	gitIndex, err := g.repo.Storer.Index()
-	if err != nil {
-		return fmt.Errorf("failed to open git index: %w", err)
-	}
-
-	// if we need to walk a path that is not the root of the repository, we will read the directory structure of the
-	// git index into memory for faster lookups
-	var idxCache *filetree
-
-	path := filepath.Clean(filepath.Join(g.root, g.path))
-	if !strings.HasPrefix(path, g.root) {
-		return fmt.Errorf("path '%s' is outside of the root '%s'", path, g.root)
-	}
-
-	switch path {
-
-	case g.root:
-
-		// we can just iterate the index entries
-		for _, entry := range gitIndex.Entries {
+	if g.scanner == nil {
+		// create a pipe to capture the command output
+		r, w := io.Pipe()
 
-			// we only want regular files, not directories or symlinks
-			if entry.Mode == filemode.Dir || entry.Mode == filemode.Symlink {
-				continue
-			}
-
-			// stat the file
-			path := filepath.Join(g.root, entry.Name)
+		// create a command which will execute from the specified sub path within root
+		cmd := exec.Command("git", "ls-files")
```

Comment:
For the record, I'm reviewing this code on my phone in bed. I haven't actually checked if ls-files does what we want when run in a subdir.

Reply:
Do we have a test of using the git walker in a subdir?

Reply:
I'll add one if there isn't